gateway/src/resty/concurrent/timer_task.lua

```diff
@@ -60,25 +56,32 @@ end
 
 local run_periodic, schedule_next, timer_execute
 
-run_periodic = function(self, run_now)
-  if not _M.task_is_active(self.id) then return end
+run_periodic = function(run_now, id, func, args, interval)
+  local t_start = ngx_now()
+
+  if not _M.task_is_active(id) then return end
 
   if run_now then
-    self.task(unpack(self.args))
+    func(unpack(args))
   end
 
-  schedule_next(self)
+  local t_diff = ngx_now() - t_start
+
+  local delay_for_next = interval - t_diff
+  if delay_for_next < 0 then delay_for_next = 0 end
```

Comment:
I think LuaJIT recommends using `math.max` instead of conditions (as they can't be  JITed).

http://wiki.luajit.org/Numerical-Computing-Performance-Guide